### PR TITLE
Nin/develop integration solv

### PR DESCRIPTION
**Pull request by SandroBui**

### CODE REVIEW

File: contracts/vaults/solv/SolvVault.sol [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+    function getLatestToken(address _token) external view returns(uint256) {
+        uint256[] memory arrToken = tokensOfOwner(address(_token));
+        uint256 latestToken = arrToken[arrToken.length - 1];
+        return latestToken;
+    }
+
+    /**
+     * @notice get total value locked vault => devide 1e26 = value real
+     */
+    function totalValueLock() external returns (uint256) {
```
Comment (baonguyen1904) [inline contracts/vaults/solv/SolvVault.sol:229]: change name to `totalValueLocked` to keep the abi consistent with others

File: contracts/vaults/solv/SolvVault.sol [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+        return (this.getPricePerShares() * vaultState.totalShares);
+    }
+
+    function getPricePerShares() external returns (uint256) {
```
Comment (baonguyen1904) [inline contracts/vaults/solv/SolvVault.sol:233]: change to `pricePerShare`

File: contracts/vaults/solv/SolvVault.sol [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    );
+
+    event Claim(
+        address indexed to,
+        uint256 indexed tokenId,
+        uint256 claimValue,
+        address currency
+    );
+
+    event requestWithdrawal(address user, address asset, uint256 shares);
```
Comment (baonguyen1904) [inline contracts/vaults/solv/SolvVault.sol:57]: please check other vault, keep the event name consistent
we need it for backend service

File: contracts/vaults/solv/SolvVault.sol [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+    event WithDrawal(address user, address asset, uint256 shares, uint256 amount, uint256 balanceBeforeWithdrawal, uint256 balanceAfterWithdrawal);
```
Comment (baonguyen1904) [inline contracts/vaults/solv/SolvVault.sol:58]: please check other vault, keep the event name consistent
we need it for backend service

Comment (baonguyen1904) [reply contracts/vaults/solv/SolvVault.sol:58]: typo as well em

File: contracts/vaults/solv/SolvVault.sol [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+        address _admin,
+        address _solvAddress,
+        address _wbtc,
+        address _tokenGOEFS,
+        address _tokenGOEFR,
+        string memory _poolId,
+        uint8 _decimal,
+        uint256 _minimumSupply,
+        uint256 _cap
+    ) public {
```
Comment (baonguyen1904) [inline contracts/vaults/solv/SolvVault.sol:70]: the initialize function must be
```
function initialize() public initializer{}
```

File: contracts/interfaces/Solv/INavOracal.sol [hunk 6]
```diff
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.19;
+
+interface INavOracal {
```
Comment (baonguyen1904) [inline contracts/interfaces/Solv/INavOracal.sol:4]: is this typo: Oracle?

File: contracts/interfaces/Solv/SolvStruct.sol [hunk 7]
```diff
@@ -0,0 +1,56 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.19;
+
+struct UserDepositSolv {
```
Comment (baonguyen1904) [inline contracts/interfaces/Solv/SolvStruct.sol:4]: For naming convension, I think inside the SolvStruct, we don't need to mention `UserDepositSolv`
just keep it as `UserDepositReceipt` or something em

File: contracts/vaults/solv/SolvVault.sol [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+            depositReceipts[msg.sender].tokenIdRedeem = 0;
+        }
+
+        emit Claim(address(this), openFundRedemptionId, shares, vaultParams.asset);
+    }
+
+    /**
+     *@notice withdrawal to the address user, only have amountWithdrawal can call this method
+     */
+    function withdrawal(uint256 shares) external nonReentrant {
```
Comment (baonguyen1904) [inline contracts/vaults/solv/SolvVault.sol:196]: please name it `completeWithdrawal` then the UI don't need to change much when we release this.

File: contracts/vaults/solv/SolvVault.sol [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+import "../../interfaces/Solv/SolvStruct.sol";
+import "../../lib/FullMath.sol";
+import "../../interfaces/Solv/ITokenGOEFR.sol";
+import "../../extensions/TransferHelper.sol";
+import "./structs/SolvStruct.sol";
+import "../../extensions/Utils.sol";
+import "hardhat/console.sol";
+
+contract SolvVault is RockOnyxAccessControl, ReentrancyGuardUpgradeable {
+    ISolv private SOLV;
```
Comment (SandroBui) [inline contracts/vaults/solv/SolvVault.sol:21]: Apply upgrade contract

File: contracts/vaults/solv/SolvVault.sol (hunk 2 above)
Comment (vuxbasnin) [reply contracts/vaults/solv/SolvVault.sol:233]: Done

File: contracts/vaults/solv/SolvVault.sol (hunk 1 above)
Comment (vuxbasnin) [reply contracts/vaults/solv/SolvVault.sol:229]: Done

**Comment by baonguyen1904** [comment]
Note this error as well em:

image

File: contracts/vaults/solv/SolvVault.sol (hunk 4 above)
Comment (vuxbasnin) [reply contracts/vaults/solv/SolvVault.sol:58]: Done

File: contracts/vaults/solv/SolvVault.sol (hunk 8 above)
Comment (vuxbasnin) [reply contracts/vaults/solv/SolvVault.sol:196]: Done

File: contracts/vaults/solv/SolvVault.sol (hunk 5 above)
Comment (vuxbasnin) [reply contracts/vaults/solv/SolvVault.sol:70]: Done

File: contracts/vaults/solv/SolvVault.sol (hunk 3 above)
Comment (vuxbasnin) [reply contracts/vaults/solv/SolvVault.sol:57]: Done

File: contracts/vaults/solv/SolvVault.sol (hunk 9 above)
Comment (vuxbasnin) [reply contracts/vaults/solv/SolvVault.sol:27]: Done

**Pull request closed without merging**